### ISSUE TITLE
Removing extra_config results in reboot on apply, even if plan says reboot_required=false

### ISSUE THREAD

**Issue by sofixa** [hashicorp/terraform-provider-vsphere#1004]
### Terraform Version
0.12.23

### vSphere Provider Version

- Checking for available provider plugins...
- Downloading plugin for provider "vsphere" (hashicorp/vsphere) 1.16.2...
- Downloading plugin for provider "vault" (hashicorp/vault) 2.9.0...
- Downloading plugin for provider "aws" (hashicorp/aws) 2.53.0...

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
variable "cachepp_num-cpus"  { default = 4 }
variable "cachepp_memory"    { default = 8 }
variable "cachepp_disk-size" { default = 100 }


# Resource PA2
##############

resource "vsphere_virtual_machine" "pa2-cachepp" {
  name             = "client.web.cachepp-01"
  guest_id         = "${data.vsphere_virtual_machine.pa2_template9.guest_id}"
  datastore_id     = "${data.vsphere_datastore.pa2_datastore_02.id}"
  resource_pool_id = "${data.vsphere_compute_cluster.pa2_compute_cluster.resource_pool_id}"

  num_cpus = "${var.cachepp_num-cpus}"
  memory   = "${var.cachepp_memory * 1024}"
 
  network_interface { #ADM
    network_id   = "${data.vsphere_network.pa2_adm.id}"
    adapter_type = "${var.network_interface_type}"
  }

  network_interface { #CACHE PP
    network_id   = "${data.vsphere_network.pa2_cache_preprod.id}"
    adapter_type = "${var.network_interface_type}"
  }

  network_interface { #FRONT PP
    network_id   = "${data.vsphere_network.pa2_front_preprod.id}"
    adapter_type = "${var.network_interface_type}"
  }

  disk {
    label            = "disk0.vmdk"
    size             = "${var.cachepp_disk-size}"
    eagerly_scrub    = "${var.eagerly_scrub}"
    thin_provisioned = "${var.thin_provisioned}"
  }

  lifecycle {
    ignore_changes = [clone[0].template_uuid, disk[0]]
  }
### THIS WAS REMOVED ###
  extra_config {
    "guestinfo.vmname" = "client.web.cachepp-02"
  }
###
  cpu_hot_add_enabled        = true
  memory_hot_add_enabled     = true
  wait_for_guest_net_routable = false

  clone {
    template_uuid = "${data.vsphere_virtual_machine.pa2_template9.id}"
  }

}

# Resource PA3
##############

resource "vsphere_virtual_machine" "pa3-cachepp" {
  name             = "client.web.cachepp-02"
  guest_id         = "${data.vsphere_virtual_machine.pa3_template9.guest_id}"
  datastore_id     = "${data.vsphere_datastore.pa3_datastore_02.id}"
  resource_pool_id = "${data.vsphere_compute_cluster.pa3_compute_cluster.resource_pool_id}"

  num_cpus = "${var.cachepp_num-cpus}"
  memory   = "${var.cachepp_memory * 1024}"
 
  network_interface { #ADM
    network_id   = "${data.vsphere_network.pa3_adm.id}"
    adapter_type = "${var.network_interface_type}"
  }

  network_interface { #CACHE PP
    network_id   = "${data.vsphere_network.pa3_cache_preprod.id}"
    adapter_type = "${var.network_interface_type}"
  }

   network_interface { #FRONT PP
    network_id   = "${data.vsphere_network.pa3_front_preprod.id}"
    adapter_type = "${var.network_interface_type}"
  }

  disk {
    label            = "disk0.vmdk"
    size             = "${var.cachepp_disk-size}"
    eagerly_scrub    = "${var.eagerly_scrub}"
    thin_provisioned = "${var.thin_provisioned}"
  }

  lifecycle {
    ignore_changes = [clone[0].template_uuid, disk[0]]
  }
### THIS WAS REMOVED ###
  extra_config {
    "guestinfo.vmname" = "client.web.cachepp-02"
  }
###
  cpu_hot_add_enabled        = true
  memory_hot_add_enabled     = true
  wait_for_guest_net_routable = false

  clone {
    template_uuid = "${data.vsphere_virtual_machine.pa3_template9.id}"
  }

}
```

### Expected Behavior

Removing an extra_config parameter [apparently results in a reboot](https://github.com/terraform-providers/terraform-provider-vsphere/blob/443552cb270911097e703a945e7d91645b8d824f/vsphere/virtual_machine_config_structure.go#L498), didn't know that. Don't know if that makes sense from the VMware side, but we weren't expecting this (it was a nice surprise on 62 machines simultaneously). However, the plan didn't indicate this:

```
[1m  # vsphere_virtual_machine.pa2-cachepp[0m will be updated in-place[0m[0m
[0m  [33m~[0m[0m resource "vsphere_virtual_machine" "pa3-cachepp" {
        [1m[0mboot_delay[0m[0m                              = 0
        [1m[0mboot_retry_delay[0m[0m                        = 10000
        [1m[0mboot_retry_enabled[0m[0m                      = false
        [1m[0mchange_version[0m[0m                          = "2019-06-18T08:47:40.729602Z"
        [1m[0mcpu_hot_add_enabled[0m[0m                     = true
        [1m[0mcpu_hot_remove_enabled[0m[0m                  = false
        [1m[0mcpu_limit[0m[0m                               = -1
        [1m[0mcpu_performance_counters_enabled[0m[0m        = false
        [1m[0mcpu_reservation[0m[0m                         = 0
        [1m[0mcpu_share_count[0m[0m                         = 4000
        [1m[0mcpu_share_level[0m[0m                         = "normal"
        [1m[0mcustom_attributes[0m[0m                       = {}
        [1m[0mdatastore_id[0m[0m                            = "datastore-69"
        [1m[0mdefault_ip_address[0m[0m                      = "10.9.77.1"
        [1m[0mefi_secure_boot_enabled[0m[0m                 = false
        [1m[0menable_disk_uuid[0m[0m                        = false
        [1m[0menable_logging[0m[0m                          = false
        [1m[0mept_rvi_mode[0m[0m                            = "automatic"
      [33m~[0m [0m[1m[0mextra_config[0m[0m                            = {
          [31m-[0m [0m"guestinfo.vmname" = "client.web.cachepp-02" [90m->[0m [0m[90mnull[0m[0m
        }
        [1m[0mfirmware[0m[0m                                = "bios"
        [1m[0mforce_power_off[0m[0m                         = true
        [1m[0mguest_id[0m[0m                                = "debian9_64Guest"
        [1m[0mguest_ip_addresses[0m[0m                      = [
            "10.72.36.108",
            "10.9.77.1",
            "10.9.78.3",
        ]
        [1m[0mhost_system_id[0m[0m                          = "host-135"
        [1m[0mhv_mode[0m[0m                                 = "hvAuto"
        [1m[0mid[0m[0m                                      = "423e8258-0e46-b4ab-3005-67cc69586193"
        [1m[0mlatency_sensitivity[0m[0m                     = "normal"
        [1m[0mmemory[0m[0m                                  = 8192
        [1m[0mmemory_hot_add_enabled[0m[0m                  = true
        [1m[0mmemory_limit[0m[0m                            = -1
        [1m[0mmemory_reservation[0m[0m                      = 0
        [1m[0mmemory_share_count[0m[0m                      = 81920
        [1m[0mmemory_share_level[0m[0m                      = "normal"
        [1m[0mmigrate_wait_timeout[0m[0m                    = 30
        [1m[0mmoid[0m[0m                                    = "vm-156"
        [1m[0mname[0m[0m                                    = "client.web.cachepp-02"
        [1m[0mnested_hv_enabled[0m[0m                       = false
        [1m[0mnum_cores_per_socket[0m[0m                    = 1
        [1m[0mnum_cpus[0m[0m                                = 4
        [1m[0mreboot_required[0m[0m                         = false
        [1m[0mresource_pool_id[0m[0m                        = "resgroup-13"
        [1m[0mrun_tools_scripts_after_power_on[0m[0m        = true
        [1m[0mrun_tools_scripts_after_resume[0m[0m          = true
        [1m[0mrun_tools_scripts_before_guest_reboot[0m[0m   = false
        [1m[0mrun_tools_scripts_before_guest_shutdown[0m[0m = true
        [1m[0mrun_tools_scripts_before_guest_standby[0m[0m  = true
        [1m[0mscsi_bus_sharing[0m[0m                        = "noSharing"
        [1m[0mscsi_controller_count[0m[0m                   = 1
        [1m[0mscsi_type[0m[0m                               = "pvscsi"
        [1m[0mshutdown_wait_timeout[0m[0m                   = 3
        [1m[0mswap_placement_policy[0m[0m                   = "inherit"
        [1m[0msync_time_with_host[0m[0m                     = false
        [1m[0mtags[0m[0m                                    = []
        [1m[0muuid[0m[0m                                    = "423e8258-0e46-b4ab-3005-67cc69586193"
        [1m[0mvapp_transport[0m[0m                          = []
        [1m[0mvmware_tools_status[0m[0m                     = "guestToolsRunning"
        [1m[0mvmx_path[0m[0m                                = "client.web.cachepp-02/client.web.cachepp-02.vmx"
        [1m[0mwait_for_guest_ip_timeout[0m[0m               = 0
        [1m[0mwait_for_guest_net_routable[0m[0m             = false
        [1m[0mwait_for_guest_net_timeout[0m[0m              = 5

        clone {
            [1m[0mlinked_clone[0m[0m  = false
            [1m[0mtemplate_uuid[0m[0m = "423eda44-3483-2160-395a-3c6173ac6bc1"
            [1m[0mtimeout[0m[0m       = 30
        }

        disk {
            [1m[0mattach[0m[0m           = false
            [1m[0mdatastore_id[0m[0m     = "datastore-69"
            [1m[0mdevice_address[0m[0m   = "scsi:0:0"
            [1m[0mdisk_mode[0m[0m        = "persistent"
            [1m[0mdisk_sharing[0m[0m     = "sharingNone"
            [1m[0meagerly_scrub[0m[0m    = true
            [1m[0mio_limit[0m[0m         = -1
            [1m[0mio_reservation[0m[0m   = 0
            [1m[0mio_share_count[0m[0m   = 1000
            [1m[0mio_share_level[0m[0m   = "normal"
            [1m[0mkeep_on_remove[0m[0m   = false
            [1m[0mkey[0m[0m              = 2000
            [1m[0mlabel[0m[0m            = "disk0.vmdk"
            [1m[0mpath[0m[0m             = "client.web.cachepp-02/client.web.cachepp-02.vmdk"
            [1m[0msize[0m[0m             = 100
            [1m[0mthin_provisioned[0m[0m = false
            [1m[0munit_number[0m[0m      = 0
            [1m[0muuid[0m[0m             = "6000C293-6d5e-f2f0-9909-e628bfef4653"
            [1m[0mwrite_through[0m[0m    = false
        }

        network_interface {
            [1m[0madapter_type[0m[0m          = "vmxnet3"
            [1m[0mbandwidth_limit[0m[0m       = -1
            [1m[0mbandwidth_reservation[0m[0m = 0
            [1m[0mbandwidth_share_count[0m[0m = 50
            [1m[0mbandwidth_share_level[0m[0m = "normal"
            [1m[0mdevice_address[0m[0m        = "pci:0:7"
            [1m[0mkey[0m[0m                   = 4000
            [1m[0mmac_address[0m[0m           = "00:50:56:be:3b:c7"
            [1m[0mnetwork_id[0m[0m            = "network-22"
            [1m[0muse_static_mac[0m[0m        = false
        }
        network_interface {
            [1m[0madapter_type[0m[0m          = "vmxnet3"
            [1m[0mbandwidth_limit[0m[0m       = -1
            [1m[0mbandwidth_reservation[0m[0m = 0
            [1m[0mbandwidth_share_count[0m[0m = 50
            [1m[0mbandwidth_share_level[0m[0m = "normal"
            [1m[0mdevice_address[0m[0m        = "pci:0:8"
            [1m[0mkey[0m[0m                   = 4001
            [1m[0mmac_address[0m[0m           = "00:50:56:be:05:73"
            [1m[0mnetwork_id[0m[0m            = "network-109"
            [1m[0muse_static_mac[0m[0m        = false
        }
        network_interface {
            [1m[0madapter_type[0m[0m          = "vmxnet3"
            [1m[0mbandwidth_limit[0m[0m       = -1
            [1m[0mbandwidth_reservation[0m[0m = 0
            [1m[0mbandwidth_share_count[0m[0m = 50
            [1m[0mbandwidth_share_level[0m[0m = "normal"
            [1m[0mdevice_address[0m[0m        = "pci:0:9"
            [1m[0mkey[0m[0m                   = 4002
            [1m[0mmac_address[0m[0m           = "00:50:56:be:2e:ee"
            [1m[0mnetwork_id[0m[0m            = "network-125"
            [1m[0muse_static_mac[0m[0m        = false
        }
    }

```

(sorry for the random characters around, it's ran through Gitlab-CI with colours).

Note the reboot_required = Fasle.

### Actual Behavior

`reboot_required=True` in cases where a VM resource needs to be rebooted.

### Steps to Reproduce

1. Create a VM with extra_config guestinfo something
2. Remove the extra_config bit
3. Plan,  `reboot_required` should be false
4. Apply, reboot will happen anyways

Linked issues:
#518 #519 

Cheers,
Adrian

**Comment by bill-rich** [comment]
Thanks! That certainly needs to be looked at. I'll get this on the backlog.

**Comment by bradphipps** [comment]
Also saw this for in-place update we did to change hostname in cloud-init. Can see `reboot_required = false` in the plan below too.

```
$ terraform --version
Terraform v0.12.23
```
Plan:
```
   # module.dev_vm_vigneshu.vsphere_virtual_machine.vm will be updated in-place
   ~ resource "vsphere_virtual_machine" "vm" {
         annotation                              = "asdasdasd."
         boot_delay                              = 0
         boot_retry_delay                        = 10000
         boot_retry_enabled                      = false
         change_version                          = "2020-02-20T18:44:45.391664Z"
         cpu_hot_add_enabled                     = true
         cpu_hot_remove_enabled                  = false
         cpu_limit                               = -1
         cpu_performance_counters_enabled        = false
         cpu_reservation                         = 0
         cpu_share_count                         = 4000
         cpu_share_level                         = "normal"
         custom_attributes                       = {}
         datastore_cluster_id                    = "group-p118"
         datastore_id                            = "datastore-658"
         default_ip_address                      = "xxx.xxx.xxx.xxx"
         efi_secure_boot_enabled                 = false
         enable_disk_uuid                        = false
         enable_logging                          = true
         ept_rvi_mode                            = "automatic"
       ~ extra_config                            = {
           ~ "guestinfo.metadata"          = jsonencode(
               ~ {
                   ~ local-hostname = "vigneshu" -> "dev-vm-vigneshu-1.desktops.speechmatics.io"
                 }
             )
             "guestinfo.userdata"          = "adasdsadads"
             "guestinfo.userdata.encoding" = "gzip+base64"
         }
         firmware                                = "bios"
         folder                                  = "foo"
         force_power_off                         = true
         guest_id                                = "ubuntu64Guest"
         guest_ip_addresses                      = [
             "10.1.14.1",
             "fe80::a86d:9aff:feae:f540",
             "10.1.14.1",
             "fe80::a86d:9aff:feae:f540",
             "192.168.7.59",
             "fe80::250:56ff:feb9:6ed",
         ]
         hardware_version                        = 9
         host_system_id                          = "host-656"
         hv_mode                                 = "hvAuto"
         id                                      = "42396f62-5168-f7fc-0607-c63ab356523b"
         latency_sensitivity                     = "normal"
         memory                                  = 16384
         memory_hot_add_enabled                  = true
         memory_limit                            = -1
         memory_reservation                      = 0
         memory_share_count                      = 163840
         memory_share_level                      = "normal"
         migrate_wait_timeout                    = 30
         moid                                    = "vm-2576"
       ~ name                                    = "foo" -> "foo.fqdn"
         nested_hv_enabled                       = false
         num_cores_per_socket                    = 1
         num_cpus                                = 4
       + poweron_timeout                         = 300
         reboot_required                         = false
         resource_pool_id                        = "resgroup-2982"
         run_tools_scripts_after_power_on        = true
         run_tools_scripts_after_resume          = true
         run_tools_scripts_before_guest_reboot   = false
         run_tools_scripts_before_guest_shutdown = true
         run_tools_scripts_before_guest_standby  = true
         scsi_bus_sharing                        = "noSharing"
         scsi_controller_count                   = 1
         scsi_type                               = "lsilogic"
         shutdown_wait_timeout                   = 3
         swap_placement_policy                   = "inherit"
         sync_time_with_host                     = true
         tags                                    = [
             "urn:vmomi:InventoryServiceTag:5d1755db-fadc-4298-b3a5-9b9ac54d46b1:GLOBAL",
         ]
         uuid                                    = "42396f62-5168-f7fc-0607-c63ab356523b"
         vapp_transport                          = []
         vmware_tools_status                     = "guestToolsRunning"
         vmx_path                                = "foo/bar.vmx"
         wait_for_guest_ip_timeout               = 0
         wait_for_guest_net_routable             = true
         wait_for_guest_net_timeout              = 5
         clone {
             linked_clone    = false
             ovf_network_map = {}
             ovf_storage_map = {}
             template_uuid   = "5e4a7cc5-9a7e-4f41-b748-487a8020743f"
             timeout         = 30
         }
         disk {
             attach           = false
             datastore_id     = "datastore-658"
             device_address   = "scsi:0:0"
             disk_mode        = "persistent"
             disk_sharing     = "sharingNone"
             eagerly_scrub    = false
             io_limit         = -1
             io_reservation   = 0
             io_share_count   = 1000
             io_share_level   = "normal"
             keep_on_remove   = false
             key              = 2000
             label            = "disk0"
             path             = "foo/bar.vmdk"
             size             = 150
             thin_provisioned = false
             unit_number      = 0
             uuid             = "6000C296-64a4-20d8-1a6b-f76b74687897"
             write_through    = false
         }
         disk {
             attach           = false
             datastore_id     = "datastore-658"
             device_address   = "scsi:0:1"
             disk_mode        = "persistent"
             disk_sharing     = "sharingNone"
             eagerly_scrub    = false
             io_limit         = -1
             io_reservation   = 0
             io_share_count   = 1000
             io_share_level   = "normal"
             keep_on_remove   = false
             key              = 2001
             label            = "disk1"
             path             = "foo/bar.vmdk"
             size             = 100
             thin_provisioned = true
             unit_number      = 1
             uuid             = "6000C29b-0f4b-ac50-bc2c-de0eeb16cdd7"
             write_through    = false
         }
         network_interface {
             adapter_type          = "e1000"
             bandwidth_limit       = -1
             bandwidth_reservation = 0
             bandwidth_share_count = 50
             bandwidth_share_level = "normal"
             device_address        = "pci:0:7"
             key                   = 4000
             mac_address           = "00:50:56:b9:06:ed"
             network_id            = "network-14"
             use_static_mac        = false
         }
     }
```

Not sure whether `poweron_timeout` is a clue?

extra_config section in VM resource looks like this:

```
  extra_config = {
    "guestinfo.userdata"          = "${data.template_cloudinit_config.cloud-config.rendered}"
    "guestinfo.userdata.encoding" = "gzip+base64"
    "guestinfo.metadata"          = <<-EOT
    {
      "local-hostname": "${var.vm_hostname}.${var.vm_domain}"
    }
    EOT
  }
```

apply also failed with this message for every VM with extra_config change, but only after several minutes. This left the VMs shut down:

```
 module.foo.vsphere_virtual_machine.vm: Still modifying... [id=42396f62-5168-f7fc-0607-c63ab356523b, 3m0s elapsed]
 Error: The attempted operation cannot be performed in the current state (Powered on).
   on ../modules/vmware-base-vm/main.tf line 64, in resource "vsphere_virtual_machine" "vm":
   64: resource "vsphere_virtual_machine" "vm" {
```

We don't want to push lots of configuration into Ansible that is run after TF, but this bug might force us to take out the cloud-init based configuration even though this is [the recommended method for configuring VMs over provisioners](https://www.terraform.io/docs/provisioners/index.html#passing-data-into-virtual-machines-and-other-compute-resources).

Willing to test any changes!

**Comment by AliAllomani** [comment]
The provider resets this value with each read, which occurs before the plan compression on state refresh; therefore i would assume you will always see it as `false` and i don't think this value should be used by user to determine if the vm is going to be restarted or not

https://github.com/hashicorp/terraform-provider-vsphere/blob/d020376d1df65b750e3ecfd764c2722fd6123934/vsphere/resource_vsphere_virtual_machine.go#L429-L431

I think docs are little bit confusing regarding this point, but it mentions that it's an internal value to terraform

https://github.com/hashicorp/terraform-provider-vsphere/blob/dd821cd16fc44301a208322780d773e95e97069f/website/docs/r/virtual_machine.html.markdown#L1608-L1610

**Comment by tenthirtyam** [comment]
Changes to, or removal of, the `extra_config` argument for the `vsphere_virtual_machine` will require a reboot.

https://github.com/hashicorp/terraform-provider-vsphere/pull/1522 by @AliAllomani addresses mentioning the reboot conditions in the `vsphere_virtual_machine` resource documentation.

To ensure that the removal of `extra_config` does not reboot the resource, use the `lifecycle` meta-argument and pass the `extra_config`.

```hcl
resource "vsphere_virtual_machine" "foo" {
  # ... other configuration ...
  lifecycle {
    ignore_changes = [
      extra_config,
    ]
  }
}
```

Marking this issue as resolved. If the issue persists with the solution provided and [the latest version of the provider](https://github.com/hashicorp/terraform-provider-vsphere/releases/latest), please open a new issue and reference this issue for additional context.

**Ryan Johnson**
Staff II Solutions Architect, VMware, Inc.
